### PR TITLE
Fix duplicate types exception on WinRT interfaces

### PR DESCRIPTION
**Pull request by OmarTawfik**
For certain WinRT types like `Windows.Foundation.Collections.IPropertySet`, when using the dynamic C# binder and calling a method, for example `Add(key, value)`, the binder creates the type hierarchy and tries to bind the base method.
The existing code tries to bind implemented interfaces both in the AggregateType's t ype hierarchy and  WinRT interfaces, and ends up with duplicates, which results in:

```Microsoft.CSharp.RuntimeBinder.RuntimeBinderException:The call is ambiguous between the following methods or properties: 'System.Collections.Generic.IDictionary<string,object>.Add(string, object)' and 'System.Collections.Generic.IDictionary<string,object>.Add(string, object)'```

The simple workaround for now is to use a `HashSet` to store these interfaces, so that we don't end up passing duplicates up the call chain.
A smarter way of filtering out types that we know for sure are not applicable should be considered in the future.

DevDiv #476163

### CODE REVIEW

**Comment by JonHanna** [comment]
Seems sound, assuming testing doesn't find some problem.

**Comment by OmarTawfik** [comment]
@dotnet-bot test this please
// https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_uap+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/2455/

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/RuntimeBinder.cs [hunk 1]
```diff
@@ -635,15 +635,17 @@ private Expr CreateArgumentEXPR(ArgumentObject argument, LocalVariableSymbol loc
...
             // into the list of candidates, and only the type containing the base
             // virtual method is inserted into the list of types. However, since we
             // don't want to do all the logic, we're just going to insert every type
             // that has a member of the given name, and allow the C# binder to filter
             // out all overrides.
-            //
+            
+            // CONSIDER: using a hashset to filter out duplicate interface types.
+            // Adopt a smarter algorithm to filter types before creating the exception.
+            HashSet<CType> callingTypes = new HashSet<CType>();
```
Comment (VSadov) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/RuntimeBinder.cs]: Could the order of types in the collection matter?

**Comment by OmarTawfik** [comment]
@dotnet-bot test UWP CoreCLR x64 Debug Build

File: src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/RuntimeBinder.cs (hunk 1 above)
Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/RuntimeBinder.cs]: @VSadov not from what I can see.

Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/RuntimeBinder.cs]: @VSadov not from what I can see, but I'll keep the existing order to make sure we're not regressing any untested scenarios. This look up needs to be re-designed afterwards. I left a comment.

Comment (OmarTawfik) [inline src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/RuntimeBinder.cs]: Upon further examination, I don't see a lot of tests for this area in particular. I'll preserve the existing order until we investigate further.

**Comment by OmarTawfik** [comment]
@VSadov comments addressed. Let me know if you want to merge this to unblock DevDiv break.

**Comment by OmarTawfik** [comment]
@dotnet-bot test Linux x64 Release Build please
https://ci3.dot.net/job/dotnet_corefx/job/master/job/linux-TGroup_netcoreapp+CGroup_Release+AGroup_x64+TestOuter_false_prtest/2584/

**Comment by JonHanna** [comment]
Are there types within corefx that would be suitable for testing this?

**Comment by OmarTawfik** [comment]
@JonHanna There are UAP apps that caught this break (in the last 2 weeks), so I need to merge this earlier to fix the break. In parallel, I'm looking into writing unit tests (here, optimally next to the dynamic binder tests) to cover that as well.

**Pull request merged**